adapter/coupling_functions.h

```diff
@@ -0,0 +1,499 @@
+     *             after each time step and exchanged with other participants.
+     * @param[in]  computed_timestep_length Length of the timestep used by
+     *             the solver.
+     */
+    void
+    advance_precice(const VectorType &deal_to_precice,
+                    VectorType &      precice_to_deal,
+                    const double      computed_timestep_length);
+
+    /**
+     * @brief      Saves current state of time dependent variables in case of an
+     *             implicit coupling
+     *
+     * @param[in]  state_variables Vector containing all variables to store as
+     *             reference
+     *
+     * @note       This function only makes sense, if it is used with
+     *             @p reload_old_state_if_required. Therefore, the order, in which the
+     *             variables are passed into the vector must be the same for
+     *             both functions.
+     * @note       The absolute time has no impact on the computation, but on the output.
+     *             Therefore, we call here in the @p Time class a method to store the
+     *             current time and reload it later. This is necessary, in
+     *             case your solver is subcycling.
+     */
+    void
+    save_current_state_if_required(
+      const std::vector<VectorType *> &state_variables,
+      Time &                           time_class);
+
+    /**
+     * @brief      Reloads the previously stored variables in case of an implicit
+     *             coupling. The current implementation supports subcycling,
+     *             i.e. previously refers o the last time
+     *             @p save_current_state_if_required() has been called.
+     *
+     * @param[out] state_variables Vector containing all variables to reload
+     *             as reference
+     *
+     * @note       This function only makes sense, if the state variables have been
+     *             stored by calling @p save_current_state_if_required. Therefore,
+     *             the order, in which the variables are passed into the
+     *             vector must be the same for both functions.
+     */
+    void
+    reload_old_state_if_required(std::vector<VectorType *> &state_variables,
+                                 Time &                     time_class);
+
+    /**
+     * @brief public precice solverinterface
+     */
+
+    precice::SolverInterface precice;
+
+    // Boundary ID of the deal.II mesh, associated with the coupling
+    // interface. The variable is public and should be used during grid
+    // generation, but is also involved during system assembly.
+    // It has been chosen arbitrarily to be 6, but it could be any other
+    // number as well. The only thing, one needs to make sure is, that this ID
+    // is not given to another part of the boundary e.g. clamped one, but this
+    // is asserted in the make_grid function
+    static constexpr unsigned int deal_boundary_interface_id = 6;
```

Comment:
Maybe you have already answered this before, but why not making this ID a parameter and giving the value in the constructor of the adapter? Then we have to modify a potential new dealii code we want to couple even less.

Reply:
Yes, I have. This is the boundary identifier, you can think of it as a patch name in OpenFOAM. Hence, it needs to be set anyway in the code. However, thinking of it again, I like the constructor idea, since it forces the user to specify it manually from his side instead looking for it in the adapter.